pd/src/verify.rs

```diff
@@ -0,0 +1,169 @@
+        let mut spend_proof_details = Vec::<SpendProofDetail>::new();
+
+        for action in self.transaction_body().actions {
+            match action {
+                Action::Output(inner) => {
+                    if !inner.body.proof.verify(
+                        inner.body.value_commitment,
+                        inner.body.note_commitment,
+                        inner.body.ephemeral_key,
+                    ) {
+                        return Err(anyhow::anyhow!("An output proof did not verify"));
+                    }
+
+                    new_notes.insert(
+                        inner.body.note_commitment,
+                        NoteData {
+                            ephemeral_key: inner.body.ephemeral_key,
+                            encrypted_note: inner.body.encrypted_note,
+                            transaction_id: id,
+                        },
+                    );
+                }
+                Action::Spend(inner) => {
+                    if !inner.verify_auth_sig() {
+                        return Err(anyhow::anyhow!("A spend auth sig did not verify"));
+                    }
+
+                    // We save this data and verify later as it requires the `App` NCT state.
+                    spend_proof_details.push(SpendProofDetail {
+                        proof: inner.body.proof,
+                        value_commitment: inner.body.value_commitment,
+                        nullifier: inner.body.nullifier.clone(),
+                        rk: inner.body.rk,
+                    });
+
+                    // Check nullifier has not been revealed already in this transaction.
+                    if spent_nullifiers.contains(&inner.body.nullifier.clone()) {
+                        return Err(anyhow::anyhow!("Double spend"));
+                    }
+
+                    spent_nullifiers.insert(inner.body.nullifier.clone());
+                }
+            }
+        }
+
+        Ok(PendingTransaction {
+            id,
+            root: self.transaction_body().merkle_root,
+            new_notes,
+            spent_nullifiers,
+            spend_proof_details,
+        })
+    }
+}
+
+impl StatefulTransactionExt for PendingTransaction {
+    fn verify_stateful(&self, nct_root: merkle::Root) -> Result<VerifiedTransaction, Error> {
+        for spend in &self.spend_proof_details {
+            if !spend.proof.verify(
+                nct_root.clone(),
+                spend.value_commitment,
+                spend.nullifier.clone(),
+                spend.rk,
+            ) {
+                return Err(anyhow::anyhow!("Spend proof does not verify"));
+            }
+        }
+
+        Ok(VerifiedTransaction {
+            id: self.id,
+            new_notes: self.new_notes.clone(),
+            spent_nullifiers: self.spent_nullifiers.clone(),
+        })
+    }
+}
+
+/// One-off function used to generate a verified genesis transaction.
```

Comment:
Maybe this could have a comment / name change to indicate that it doesn't actually do verification? Like `mark_genesis_as_verified` or something?